src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/PropertySetAnalysis/PropertySetCallbacks.cs

```diff
@@ -0,0 +1,177 @@
+                    {
+                        // We know all values but some values are bad, so we can say MaybeFlagged.
+                        return PropertySetAbstractValueKind.MaybeFlagged;
+                    }
+                    else
+                    {
+                        // We know all values are good, so we can say Unflagged.
+                        return PropertySetAbstractValueKind.Unflagged;
+                    }
+
+                case ValueContainsNonLiteralState.Maybe:
+                    if (valueContentAbstractValue.LiteralValues.Any(badLiteralValuePredicate))
+                    {
+                        // We don't know all values but know some values are bad, so we can say MaybeFlagged.
+                        return PropertySetAbstractValueKind.MaybeFlagged;
+                    }
+                    else
+                    {
+                        // We don't know all values but didn't find any bad value, so we can say who knows.
+                        return PropertySetAbstractValueKind.Unknown;
+                    }
+
+                default:
+                    return PropertySetAbstractValueKind.Unknown;
+            }
+        }
+
+        /// <summary>
+        /// A <see cref="HazardousUsageEvaluator.EvaluationCallback"/> for all properties flagged being hazardous.
+        /// </summary>
+        /// <param name="propertySetAbstractValue">PropertySetAbstract value.</param>
+        /// <returns>If all properties are flagged, then flagged; if at least one property is unflagged, then unflagged; otherwise maybe flagged.</returns>
+        public static HazardousUsageEvaluationResult HazardousIfAllFlagged(PropertySetAbstractValue propertySetAbstractValue)
+        {
+            if (propertySetAbstractValue.KnownValuesCount == 0)
+            {
+                // No known values implies only PropertySetAbstractValueKind.Unknown.
+                return HazardousUsageEvaluationResult.MaybeFlagged;
+            }
+
+            bool allFlagged = true;
+            bool atLeastOneUnflagged = false;
+            for (int i = 0; i < propertySetAbstractValue.KnownValuesCount; i++)
+            {
+                if (propertySetAbstractValue[i] != PropertySetAbstractValueKind.Flagged)
+                {
+                    allFlagged = false;
+                }
+
+                if (propertySetAbstractValue[i] == PropertySetAbstractValueKind.Unflagged)
+                {
+                    atLeastOneUnflagged = true;
+                    break;
+                }
+            }
+
+            if (allFlagged)
+            {
+                return HazardousUsageEvaluationResult.Flagged;
+            }
+            else if (atLeastOneUnflagged)
+            {
+                return HazardousUsageEvaluationResult.Unflagged;
+            }
+            else
+            {
+                return HazardousUsageEvaluationResult.MaybeFlagged;
+            }
```

Comment:
Also need to deal with Unknown? Otherwise, for property that is all unknown, which is passed by parameter, it will be flagged as MaybeFlagged.

Reply:
That's by design (so far).  All unknown is handled above by the first if statement.  I think that makes sense for scenarios where the defaults are insecure (like the BinaryFormatter and NetDataContractSerializer rules).

But for scenarios where the defaults are secure, then we probably should treat all unknowns as unflagged, so we should create a different callback when we need it.

Have any good ideas for method names?

---
In reply to: [296162589](https://github.com/dotnet/roslyn-analyzers/pull/2415#discussion_r296162589) [](ancestors = 296162589)

Reply:
And Json.NET is a scenario where the defaults are secure, so I'll make a new one.

While writing unit tests for an all unknown case, ran into some issues with my current approach, so I have more work to do.

---
In reply to: [296376529](https://github.com/dotnet/roslyn-analyzers/pull/2415#discussion_r296376529) [](ancestors = 296376529,296162589)